cmd/kindnetd/main.go

```diff
@@ -0,0 +1,198 @@
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// obtain the host and pod ip addresses
+	// if both ips are different we are not using the host network
+	hostIP, podIP := os.Getenv("HOST_IP"), os.Getenv("POD_IP")
+	fmt.Printf("hostIP = %s\npodIP = %s\n", hostIP, podIP)
+	if hostIP != podIP {
+		panic(fmt.Sprintf(
+			"hostIP(= %q) != podIP(= %q) but must be running with host network: ",
+			hostIP, podIP,
+		))
+	}
+
+	// used to track if the cni config inputs changed
+	lastCNIData := cniConfigInputs{}
+
+	// main control loop
+	for {
+		// Gets the Nodes information from the API
+		nodes, err := clientset.CoreV1().Nodes().List(metav1.ListOptions{})
+		if err != nil {
+			panic(err.Error())
+		}
+		fmt.Printf("There are %d nodes in the cluster\n", len(nodes.Items))
+
+		// Iterate over all the nodes information
+		for _, node := range nodes.Items {
+			// Obtain node internal IP
+			nodeIP := internalIP(node)
+			fmt.Printf("Handling node with IP: %s\n", nodeIP)
+
+			// this is our node, handle it specially
+			// we don't need to add routes, but we might need to update
+			// the cni config
+			if nodeIP == hostIP {
+				fmt.Printf("handling current node\n")
+				// compute the current cni config inputs
+				cniData := computeCNIConfigInputs(node)
+				// if it changed, write out the new config
+				if cniData != lastCNIData {
+					fmt.Printf("Writing CNI data\n")
+					if err := writeCNIConfig(cniData); err != nil {
+						panic(err.Error())
+					}
+					lastCNIData = cniData
+				}
+				// we're don handling this node
+				continue
+			}
+
+			// Obtain Pod Subnet
+			if node.Spec.PodCIDR == "" {
+				fmt.Printf("Node %v has no CIDR, ignoring\n", node.Name)
+				continue
+			}
+
+			dst, err := netlink.ParseIPNet(node.Spec.PodCIDR)
+			if err != nil {
+				panic(err.Error())
+			}
+			fmt.Printf("Node %v has CIDR %s \n",
+				node.Name, node.Spec.PodCIDR)
+
+			// Check if the route exists
+			ip := net.ParseIP(nodeIP)
+			routeToDst := netlink.Route{Dst: dst, Gw: ip}
+			route, err := netlink.RouteListFiltered(nl.GetIPFamily(ip), &routeToDst, netlink.RT_FILTER_DST)
+			if err != nil {
+				panic(err.Error())
+			}
+
+			// Add route if not present
+			if len(route) == 0 {
+				if err := netlink.RouteAdd(&routeToDst); err != nil {
+					panic(err.Error())
+				}
+				fmt.Printf("Adding route to the system %v \n", routeToDst)
+			}
+		}
+
+		// rate limit
+		time.Sleep(10 * time.Second)
+	}
+}
+
+func internalIP(node corev1.Node) string {
+	for _, address := range node.Status.Addresses {
+		if address.Type == "InternalIP" {
+			return address.Address
+		}
+	}
+	return ""
+}
+
+func computeCNIConfigInputs(node corev1.Node) cniConfigInputs {
+	podCIDR := node.Spec.PodCIDR
+	return cniConfigInputs{
+		PodCIDR: podCIDR,
+	}
+}
+
+func writeCNIConfig(data cniConfigInputs) error {
+	return fs.AtomicWriteFile("/etc/cni/net.d/10-kindnet.conflist", func(w io.Writer) error {
```

Comment:
this doesn't actually do what I wanted in retrospect: this will still create a file in the cni directory, instead we want to mimic what configmaps do: 
- write somewhere else and then create / update a symlink

this is to avoid spurious CNI reload, containerd (and I think cri-o) do an fsnotify watch on the cni config

Reply:
Should we bump the file number? s/10/90/ per example, this will allow other CNIs to take control without any additional changes